### DISCUSSION TITLE
HA sensor as input in automation script

### DISCUSSION THREAD

**Discussion by Ric57100** [nathanmarlor/foxess_modbus#560] [Ideas]
Hi!

First of all, thank you for this integration! Previously I used it with Direct LAN mode, and recently with Modbus/Serial to Ethernet/WiFi adapters (Elfin EW11).

I really appreciate the ability to set forced load periods on grid, but it does not seem possible to set the periods by HA variable.

Let me explain. I recently changed my contract with my supplier to switch to "Tempo" offer. EDF's Tempo offer is a type of electricity pricing that varies according to consumption periods and days of the week. Here's how the Tempo offer typically works: 

**Three types of days:**
- Blue Days: These are the days when electricity is cheapest, representing about 300 days per year.
- White Days: Electricity is slightly more expensive than on Blue Days but cheaper than on Red Days, representing 40 days per year.
- Red Days: These are the days when electricity is most expensive, representing 25 days per year, from November 1 to March 31 excluding weekends and statutory holidays and often corresponding to periods of high demand, such as very cold days.

**Two differentiated time slots during the day :** 
- Off-peak hours, from 22:00 to 06:00, when electricity is cheaper.
- Peak hours, from 06:00 to 22:00, when electricity is more expensive.

With all that, it gives me a total of 6 different tariffs.

Given that my inverter has a battery pack, I am trying to optimize their charge according to the next day's price. The formula _(Total Battery capacity - Remaining Battery Capacity) / Charging Rate_ gives me the required time to charge my batteries. The end of the charging period is always fixed at 05:45. _End Charging Periode - Time required to charge_ give me the start of the charging period.

The integration does not seem to support HA sensor as input in an automation script like that:
service: foxess_modbus.update_charge_period
data:
  enable_force_charge: true
  enable_charge_from_grid: true
  start: **_"{{ states('sensor.start_time') }}"_**
  end: "05:45:00"
  charge_period: "2"

Am I wrong or do you confirm? Would it be possible for you to make this adaptation in the near future?

Thank a lot for the work already done.

**Comment by FozzieUK** [comment]
Yes it does support doing that, I use this a lot but you need to make sure it is in the correct format ‘HH:MM:SS’ if start_time is a datetime you will need to format it so it presents as ‘HH:MM:00’

**Reply by canton7** [reply]
I'm open to making the parsing more permissive there. If you can give me the actual value of `sensor.start_time` (use Developer Tools - Template), I can see whether it's possible to add support for parsing it.

**Comment by Ric57100** [comment]
Thanks for your replies. My sensor was time format. I just tried with a text sensor and it works.

**Reply by FozzieUK** [reply]
ok, great.
If you use a helper set to 'time' it will work fine, or if you are using a datetime you can convert on the fly something like this
```
  end: >
    {{ states('sensor.start_time')|as_timestamp|timestamp_custom("%H:%M:%S") }}
```
